adapters/rack/lib/opentelemetry/adapters/rack/middlewares/tracer_middleware.rb

```diff
@@ -0,0 +1,193 @@
+            end
+
+            def build_attribute_name(prefix, suffix)
+              prefix + suffix.to_s.downcase.gsub(/[-\s]/, '_')
+            end
+
+            def config
+              Rack::Adapter.instance.config
+            end
+
+            private
+
+            def clear_cached_config
+              @allowed_rack_request_headers = nil
+              @allowed_response_headers = nil
+            end
+          end
+
+          EMPTY_HASH = {}.freeze
+
+          def initialize(app)
+            @app = app
+          end
+
+          def call(env)
+            original_env = env.dup
+            extracted_context = OpenTelemetry.propagation.extract(env)
+            frontend_context = create_frontend_span(env, extracted_context)
+
+            # restore extracted context in this process:
+            OpenTelemetry::Context.with_current(frontend_context || extracted_context) do
+              request_span_name = create_request_span_name(env['REQUEST_URI'] || original_env['PATH_INFO'])
+              tracer.in_span(request_span_name,
+                             attributes: request_span_attributes(env: env),
+                             kind: :server) do |request_span|
+                @app.call(env).tap do |status, headers, response|
+                  set_attributes_after_request(request_span, status, headers, response)
+                end
+              end
+            end
+          ensure
+            finish_span(frontend_context)
+          end
+
+          private
+
+          # return Context with the frontend span as the current span
+          def create_frontend_span(env, extracted_context)
+            request_start_time = OpenTelemetry::Adapters::Rack::Util::QueueTime.get_request_start(env)
+
+            return unless config[:record_frontend_span] && !request_start_time.nil?
+
+            span = tracer.start_span('http_server.queue',
+                                     with_parent_context: extracted_context,
+                                     attributes: {
+                                       'component' => 'http',
+                                       'service' => config[:web_service_name],
+                                       'start_time' => request_start_time
+                                     },
+                                     kind: :server)
+
+            extracted_context.set_value(current_span_key, span)
+          end
+
+          def finish_span(context)
+            context[current_span_key]&.finish if context
+          end
+
+          def current_span_key
+            OpenTelemetry::Trace::Propagation::ContextKeys.current_span_key
+          end
+
+          def tracer
+            OpenTelemetry::Adapters::Rack::Adapter.instance.tracer
+          end
+
+          ### request_span
+
+          def request_span_attributes(env:)
+            {
+              'component' => 'http',
+              'http.method' => env['REQUEST_METHOD'],
+              'http.host' => env['HOST'],
```

Comment:
This should be:

```suggestion
              'http.host' => env['HTTP_HOST'] || 'unknown',
```

`HOST` is normalized to `HTTP_HOST` by Rack. The 'unknown' is debatable, but I have that this can be, and regularly is nil, which is not a valid attribute value. The example app will expose this:

```
I, [2020-02-24T17:59:31.953568 #89019]  INFO -- : Adapter: OpenTelemetry::Adapters::Rack was successfully installed
#<struct OpenTelemetry::SDK::Trace::SpanData
 name="/",
 kind=:server,
 status=
  #<OpenTelemetry::Trace::Status:0x00007f9562141b40
   @canonical_code=0,
   @description="">,
 parent_span_id="0000000000000000",
 child_count=0,
 total_recorded_attributes=7,
 total_recorded_events=0,
 total_recorded_links=0,
 start_timestamp=2020-02-24 17:59:31 -0800,
 end_timestamp=2020-02-24 17:59:31 -0800,
 attributes=
  {"component"=>"http",
   "http.method"=>"GET",
   "http.host"=>nil,
   "http.scheme"=>"http",
   "http.target"=>"/",
   "http.status_code"=>200,
   "http.status_text"=>"OK"},
 links=nil,
 events=nil,
 library_resource=
  #<OpenTelemetry::SDK::Resources::Resource:0x00007f95648c3a10
   @labels={"name"=>"OpenTelemetry::Adapters::Rack", "version"=>"0.0.0"}>,
 span_id="7b9ccc4a17b7355c",
 trace_id="9e6ad9993469c5443ef979de4a565bba",
 trace_flags=#<OpenTelemetry::Trace::TraceFlags:0x00007f9564888af0 @flags=1>>
#<Rack::MockResponse:0x00007f95621a83b8>
```